File: packages/neptune_flutter_ui/lib/src/widgets/neptune_status_motion.dart
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// The Odyssey outcome motion — a loading indicator for in-flight work that
// hands off smoothly to an animated SUCCESS check or an animated REJECTED
// cross. One widget, three linked states: drive it with a [NeptuneFlowStatus]
// and the transitions (spin-out → spring-in, stroke-drawn glyphs, rejection
// shake) run on the brand's motion curves. Honours reduced-motion by rendering
// static glyphs. Theme-only, RTL-safe.
//
// R6: the loading phase is no longer hourglass-only — [loaderStyle] picks
// any of the standalone loaders in neptune_loaders.dart, all sharing the same
// hand-off choreography into the outcome disc.

import 'dart:math' as math;

import 'package:flutter/material.dart';

import '../theme/extensions.dart';
import 'neptune_loaders.dart';

/// The three linked states of [NeptuneStatusMotion].
enum NeptuneFlowStatus { loading, success, rejected }

/// Animated loader → check / cross outcome indicator.
///
/// While [status] is [NeptuneFlowStatus.loading], [loaderStyle] (hourglass by
/// default) animates on a gentle loop. Flip the status to `success` and it
/// spins away while a tinted disc springs in and DRAWS the check stroke
/// ([color] defaults to the brand success role — pass any colour); `rejected`
/// draws the cross in the error role with a decaying shake.
class NeptuneStatusMotion extends StatelessWidget {
  final NeptuneFlowStatus status;

  /// Diameter of the indicator disc.
  final double size;

  /// Override the success tint (defaults to the brand `success` role).
  final Color? color;

  /// Which loader animates during [NeptuneFlowStatus.loading].
  final NeptuneLoaderStyle loaderStyle;

  const NeptuneStatusMotion({
    super.key,
    required this.status,
    this.size = 96,
    this.color,
    this.loaderStyle = NeptuneLoaderStyle.hourglass,
  });

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final motion = theme.extension<NptMotion>()!;

    // Self-size to [size]x[size] regardless of the ambient constraints — a
    // caller that asks for a 168dp indicator shouldn't also have to know to
    // wrap it in a SizedBox. Without this, AnimatedSwitcher sizes to its
    // child's own intrinsic size, and not every loaderStyle/outcome-disc
    // combination self-constrains identically in a loose-constraint context
    // (e.g. a bare Column(mainAxisSize: min) in a widget test), which showed
    // up as a five-figure RenderFlex overflow in a real consuming app.
    return SizedBox(
      width: size,
      height: size,
      child: AnimatedSwitcher(
        duration: motion.durationStandard,
        switchInCurve: motion.spring,
        switchOutCurve: motion.standard,
        transitionBuilder: (child, animation) => ScaleTransition(
          scale: animation,
          child: RotationTransition(
            // A quarter-turn hand-off links the loader to the outcome disc.
            turns: Tween<double>(begin: 0.25, end: 0).animate(animation),
            child: FadeTransition(opacity: animation, child: child),
          ),
        ),
        child: switch (status) {
          NeptuneFlowStatus.loading =>
            neptuneLoaderFor(loaderStyle, key: const ValueKey('loading'), size: size),
          NeptuneFlowStatus.success => _OutcomeDisc(
              key: const ValueKey('success'),
              size: size,
              success: true,
              tint: color,
            ),
          NeptuneFlowStatus.rejected => _OutcomeDisc(
              key: const ValueKey('rejected'),
              size: size,
              success: false,
              tint: color,
            ),
        },
      ),
    );
  }
}

// --- outcome disc (check / cross) --------------------------------------------

class _OutcomeDisc extends StatefulWidget {
  final double size;
  final bool success;
  final Color? tint;

  const _OutcomeDisc({
    super.key,
    required this.size,
    required this.success,
    this.tint,
  });

  @override
  State<_OutcomeDisc> createState() => _OutcomeDiscState();
}

class _OutcomeDiscState extends State<_OutcomeDisc>
    with SingleTickerProviderStateMixin {
  late final AnimationController _c;

  @override
  void initState() {
    super.initState();
    _c = AnimationController(vsync: this, duration: const Duration(milliseconds: 700));
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    if (MediaQuery.of(context).disableAnimations) {
      _c.value = 1;
    } else if (_c.value == 0) {
      _c.forward();
    }
  }

  @override
  void dispose() {
    _c.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final scheme = theme.colorScheme;
    final npt = theme.extension<NptColors>()!;

    final Color fill;
    final Color glyph;
    if (widget.success) {
      fill = widget.tint ?? npt.success;
      glyph = npt.onSuccess;
    } else {
      fill = widget.tint ?? scheme.error;
      glyph = scheme.onError;
    }

    return AnimatedBuilder(
      animation: _c,
      builder: (context, _) {
        final t = _c.value;
        final draw = Curves.easeOutCubic.transform(t);
        // Rejection carries a decaying shake while the cross draws.
        final shake = widget.success
            ? 0.0
            : math.sin(t * math.pi * 5) * (1 - t) * widget.size * 0.05;
        return Transform.translate(
          offset: Offset(shake, 0),
          child: Container(
            width: widget.size,
            height: widget.size,
            decoration: BoxDecoration(
              color: fill,
              shape: BoxShape.circle,
              boxShadow: [
                BoxShadow(
                  color: fill.withValues(alpha: 0.35),
                  blurRadius: widget.size * 0.25,
                  offset: Offset(0, widget.size * 0.08),
                ),
              ],
            ),
            child: CustomPaint(
              painter: _GlyphPainter(
                progress: draw,
                color: glyph,
                success: widget.success,
              ),
              size: Size.square(widget.size),
            ),
          ),
        );
      },
    );
  }
}

/// Stroke-draws the check or cross from 0 → [progress].
class _GlyphPainter extends CustomPainter {
  final double progress;
  final Color color;
  final bool success;

  const _GlyphPainter({
    required this.progress,
    required this.color,
    required this.success,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final w = size.width;
    final paint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = w * 0.09
      ..strokeCap = StrokeCap.round
      ..strokeJoin = StrokeJoin.round;

    final Path path;
    if (success) {
      path = Path()
        ..moveTo(w * 0.30, w * 0.52)
        ..lineTo(w * 0.45, w * 0.66)
        ..lineTo(w * 0.71, w * 0.36);
    } else {
      path = Path()
        ..moveTo(w * 0.35, w * 0.35)
        ..lineTo(w * 0.65, w * 0.65)
        ..moveTo(w * 0.65, w * 0.35)
        ..lineTo(w * 0.35, w * 0.65);
    }

    // Draw the stroke progressively along its metrics.
    var total = 0.0;
    final metrics = path.computeMetrics().toList();
    for (final m in metrics) {
      total += m.length;
    }
    var budget = total * progress;
    for (final m in metrics) {
      if (budget <= 0) break;
      final len = math.min(budget, m.length);
      canvas.drawPath(m.extractPath(0, len), paint);
      budget -= len;
    }
  }

  @override
  bool shouldRepaint(_GlyphPainter old) =>
      old.progress != progress || old.color != color || old.success != success;
}

File: packages/neptune_kmp_ui/odyssey-compose-ui/src/commonMain/kotlin/ly/neptune/odyssey/ui/components/NeptuneStatusMotion.kt
// © 2026 Neptune.Fintech (neptune.ly) · Neptune Odyssey Community License v1.0
//
// The Odyssey outcome motion — a loading indicator for in-flight work that
// hands off smoothly to an animated SUCCESS check or an animated REJECTED
// cross. One composable, three linked states: drive it with a
// [NeptuneFlowStatus] and the transitions (spin-out → spring-in, stroke-drawn
// glyphs, rejection shake) run on the brand's motion curves. Web counterpart:
// `<npt-status-motion state="loading|success|rejected">` · Flutter:
// neptune_status_motion.dart. Honours reduced-motion by rendering the final
// frame statically. Theme-only, RTL-safe.
//
// R6: the loading phase is no longer hourglass-only — [loaderStyle] picks any
// of the standalone loaders in NeptuneLoaders.kt, all sharing the same
// hand-off choreography into the outcome disc.

package ly.neptune.odyssey.ui.components

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.EnterExitState
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.tween
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.scaleIn
import androidx.compose.animation.scaleOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathMeasure
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.StrokeJoin
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import ly.neptune.odyssey.ui.identity.nptShadow
import ly.neptune.odyssey.ui.theme.NeptuneTheme
import ly.neptune.odyssey.ui.theme.NptShadow
import kotlin.math.PI
import kotlin.math.min
import kotlin.math.sin

// Flutter's easeOutCubic, ported numerically: cubic (0.215, 0.61, 0.355, 1).
private val easeOutCubic: CubicBezierEasing = CubicBezierEasing(0.215f, 0.61f, 0.355f, 1f)

/**
 * The three linked states of [NeptuneStatusMotion].
 *
 * Web counterpart: `<npt-status-motion state=…>` · Flutter:
 * `NeptuneFlowStatus`.
 */
public enum class NeptuneFlowStatus { Loading, Success, Rejected }

/**
 * Animated loader → check / cross outcome indicator.
 *
 * While [status] is [NeptuneFlowStatus.Loading], [loaderStyle] (hourglass by
 * default) animates on a gentle loop. Flip the status to `Success` and it
 * spins away while a tinted disc springs in and DRAWS the check stroke
 * ([color] defaults to the brand `success` role — pass any colour);
 * `Rejected` draws the cross in the `error` role with a decaying shake.
 * The hand-off is a scale × quarter-turn × fade on the brand's `spring`
 * curve in / `standard` curve out. Under reduced motion the target state
 * renders statically at its final frame.
 *
 * Web counterpart: `<npt-status-motion>` · Flutter: `NeptuneStatusMotion`.
 */
@Composable
public fun NeptuneStatusMotion(
    status: NeptuneFlowStatus,
    modifier: Modifier = Modifier,
    size: Dp = 96.dp,
    color: Color? = null,
    loaderStyle: NeptuneLoaderStyle = NeptuneLoaderStyle.Hourglass,
) {
    val motion = NeptuneTheme.motion
    val reduced = NeptuneTheme.reducedMotion

    if (reduced) {
        // No hand-off choreography — the target's final frame, statically
        // (each loader freezes itself; the disc jumps to completed t = 1).
        Box(modifier, contentAlignment = Alignment.Center) {
            StatusFrame(status, size, color, loaderStyle)
        }
        return
    }

    AnimatedContent(
        targetState = status,
        modifier = modifier,
        transitionSpec = {
            // Flutter AnimatedSwitcher: duration standard, switch-in on the
            // brand spring curve, switch-out on the brand standard curve,
            // scale × fade (the quarter turn rides the child below).
            val inSpec = tween<Float>(motion.standardMs, easing = motion.spring)
            val outSpec = tween<Float>(motion.standardMs, easing = motion.standard)
            (fadeIn(inSpec) + scaleIn(inSpec, initialScale = 0f))
                .togetherWith(fadeOut(outSpec) + scaleOut(outSpec, targetScale = 0f))
        },
        contentAlignment = Alignment.Center,
        label = "statusMotion",
    ) { target ->
        // A quarter-turn hand-off links the loader to the outcome disc:
        // incoming unwinds 90° → 0 on the spring curve, outgoing winds back
        // up on the standard curve (Flutter turns 0.25 → 0).
        val turns = transition.animateFloat(
            transitionSpec = {
                if (targetState == EnterExitState.Visible) {
                    tween(motion.standardMs, easing = motion.spring)
                } else {
                    tween(motion.standardMs, easing = motion.standard)
                }
            },
            label = "quarterTurn",
        ) { state -> if (state == EnterExitState.Visible) 0f else 90f }
        Box(Modifier.graphicsLayer { rotationZ = turns.value }) {
            StatusFrame(target, size, color, loaderStyle)
        }
    }
}

/** One frame of the status hand-off: the loader or an outcome disc. */
@Composable
private fun StatusFrame(
    status: NeptuneFlowStatus,
    size: Dp,
    color: Color?,
    loaderStyle: NeptuneLoaderStyle,
) {
    when (status) {
        NeptuneFlowStatus.Loading -> neptuneLoaderFor(loaderStyle, size = size)
        NeptuneFlowStatus.Success -> OutcomeDisc(success = true, size = size, tint = color)
        NeptuneFlowStatus.Rejected -> OutcomeDisc(success = false, size = size, tint = color)
    }
}

/**
 * The tinted outcome disc: springs in over 700ms while the check/cross is
 * STROKE-DRAWN progressively on easeOutCubic; rejection carries a decaying
 * horizontal shake. Reduced motion jumps straight to the completed frame.
 */
@Composable
private fun OutcomeDisc(
    success: Boolean,
    size: Dp,
    tint: Color?,
    modifier: Modifier = Modifier,
) {
    val scheme = MaterialTheme.colorScheme
    val npt = NeptuneTheme.colors
    val reduced = NeptuneTheme.reducedMotion
    val shape = NeptuneTheme.shape.rFull

    val fill = tint ?: if (success) npt.success else scheme.error
    val glyph = if (success) npt.onSuccess else scheme.onError

    val progress = remember(reduced) { Animatable(if (reduced) 1f else 0f) }
    if (!reduced) {
        LaunchedEffect(progress) {
            progress.animateTo(1f, tween(700, easing = LinearEasing))
        }
    }

    Canvas(
        modifier
            .graphicsLayer {
                // Rejection carries a decaying shake while the cross draws:
                // sin(t·5π) · (1 − t) · 5% of the disc diameter.
                if (!success) {
                    val t = progress.value
                    translationX = sin(t * 5f * PI.toFloat()) * (1f - t) * this.size.width * 0.05f
                }
            }
            // The disc's soft self-coloured halo (Flutter BoxShadow:
            // fill @ 35%, blur 25% of size, offset-y 8% of size).
            .nptShadow(
                listOf(
                    NptShadow(
                        color = fill.copy(alpha = 0.35f),
                        blurRadius = size * 0.25f,
                        offsetY = size * 0.08f,
                    ),
                ),
                shape,
            )
            .size(size),
    ) {
        val w = this.size.width
        drawCircle(fill, radius = this.size.minDimension / 2f)

        // Stroke-draw the glyph from 0 → progress along its contours
        // (the Flutter _GlyphPainter metrics walk).
        val draw = easeOutCubic.transform(progress.value)
        val contours: List<Path> = if (success) {
            listOf(
                Path().apply {
                    moveTo(w * 0.30f, w * 0.52f)
                    lineTo(w * 0.45f, w * 0.66f)
                    lineTo(w * 0.71f, w * 0.36f)
                },
            )
        } else {
            listOf(
                Path().apply {
                    moveTo(w * 0.35f, w * 0.35f)
                    lineTo(w * 0.65f, w * 0.65f)
                },
                Path().apply {
                    moveTo(w * 0.65f, w * 0.35f)
                    lineTo(w * 0.35f, w * 0.65f)
                },
            )
        }
        val style = Stroke(width = w * 0.09f, cap = StrokeCap.Round, join = StrokeJoin.Round)
        val measures = contours.map { path -> PathMeasure().also { it.setPath(path, false) } }
        val total = measures.fold(0f) { acc, m -> acc + m.length }
        var budget = total * draw
        for (m in measures) {
            if (budget <= 0f) break
            val len = min(budget, m.length)
            val segment = Path()
            if (m.getSegment(0f, len, segment, startWithMoveTo = true)) {
                drawPath(segment, glyph, style = style)
            }
            budget -= len
        }
    }
}
